### PR TITLE
Creating a branch to work with MapR-FS

### PR DESCRIPTION
**Pull request by stevenmanton**
We have a desire to work with dask on MapR. However, MapR has its own implementation of HDFS and does not use libhdfs3, but rather uses its own C library. The MapR API is similar to that of libhfds3, but it's missing some functions.

One key questions we want to address is whether to attempt to modify the hdfs3 package to simultaneously support both libhdfs3 and mapr, or to maintain two separate packages. The advantage of a single package is that the majority of the codebase is the same.

Core devs, please have a look and help weigh in on which option would be best moving forward.

Note: This branch is not fully working because of the API differences. Once we decide how we want the packaging, we can clean up the code and replace the missing functionality.

### CODE REVIEW

File: hdfs3/core.py [hunk 1]
```diff
@@ -207,7 +209,8 @@ def connect(self):
             #    self._token = _lib.hdfsGetDelegationToken(self._handle,
             #                                             ensure_bytes(self.user))
         else:
-            msg = ensure_string(_lib.hdfsGetLastError())
+            # msg = ensure_string(_lib.hdfsGetLastError())
```
Comment (stevenmanton) [inline hdfs3/core.py]: The MapR API does not have the function `hdfsGetLastError`

File: hdfs3/tests/test_hdfs3.py [hunk 2]
```diff
@@ -194,44 +200,46 @@ def test_replication(hdfs):
     path = '/tmp/test/afile'
     hdfs.open(path, 'wb', replication=0).close()
     assert hdfs.info(path)['replication'] > 0
-    hdfs.open(path, 'wb', replication=1).close()
-    assert hdfs.info(path)['replication'] == 1
-    hdfs.open(path, 'wb', replication=2).close()
-    assert hdfs.info(path)['replication'] == 2
-    hdfs.set_replication(path, 3)
+    # hdfs.open(path, 'wb', replication=1).close()
```
Comment (stevenmanton) [inline hdfs3/tests/test_hdfs3.py]: The MapR API doesn't support specifying replication

File: hdfs3/lib.py [hunk 3]
```diff
@@ -522,54 +523,54 @@ class hdfsFS(ct.Structure):
...
-return return 0 on success, -1 on error."""
-
-hdfsFreeFileBlockLocations = _lib.hdfsFreeFileBlockLocations
-hdfsFreeFileBlockLocations.argtypes = [ct.POINTER(BlockLocation), ct.c_int]
-hdfsFreeFileBlockLocations.restype = None
-hdfsFreeFileBlockLocations.__doc__ = """Free the BlockLocation array returned by hdfsGetFileBlockLocations
-
-param locations The array returned by hdfsGetFileBlockLocations
-param numOfBlock The number of elements in the locations"""
+# hdfsTruncate = _lib.hdfsTruncate
```
Comment (stevenmanton) [inline hdfs3/lib.py]: MapR doesn't support the functions in this block

File: hdfs3/tests/test_hdfs3.py [hunk 4]
```diff
@@ -73,7 +74,13 @@ def test_connection_error():
     # error message is long and with java exceptions, so here we just check
     # that important part of error is present
     msg = 'Caused by: HdfsNetworkConnectException: Connect to "localhost:9999"'
-    assert msg in str(ctx.value)
+    # assert msg in str(ctx.value)
```
Comment (stevenmanton) [inline hdfs3/tests/test_hdfs3.py]: MapR can't read the exception string

File: hdfs3/core.py [hunk 5]
```diff
@@ -274,6 +277,8 @@ def df(self):
 
     def get_block_locations(self, path, start=0, length=0):
         """ Fetch physical locations of blocks """
+        raise NotImplementedError()
```
Comment (stevenmanton) [inline hdfs3/core.py]: Not supported by MapR

**Comment by mrocklin** [comment]
Sorry for the delay in responding to this.  Conferences and visiting family got in the way.

Given the differences I think it makes sense to fork and produce a different library.  We could try to align the two libraries in the future if we found that maintaining both was expensive.

cc @martindurant for thoughts.

**Comment by martindurant** [comment]
One the hdfs3 code wash fleshed out earlier this year, it has not been much of a maintenance burden at all. I agree that there do seem to be enough differences between this and the mapr implementation (particularly in build dependencies) that it makes sense to have them separate. 

We do now have a number of these little libraries, including s3fs and adlfs, so when thinking about integrating them, e.g., with dask, it is in our interests to keep them similar enough to not have to do too much extra work.

**Comment by stevenmanton** [comment]
Thanks for the feedback. It seems like there's unanimous feedback to maintain this as a separate forked project. I think we will probably rename the repo and python package to something like `maprfs` and continue to work on the compatibility with dask. Thanks for the tips!

**Comment by mrocklin** [comment]
Sounds good.  Please reach out with questions about supporting this within Dask sooner rather than later.  This should be easy to extend, but there is not good reference on how to do this other than through existing sourcecode.

**Pull request closed without merging**